src/ssl.c

```diff
@@ -260,6 +262,590 @@ const WOLF_EC_NIST_NAME kNistCurves[] = {
 };
 #endif
 
+#if defined(HAVE_ECH)
+/* create the hpke key and ech config to send to clients */
+int wolfSSL_CTX_GenerateEchConfig(WOLFSSL_CTX* ctx, const char* publicName)
+{
+    int ret = 0;
+    int rngRet = 0;
+    word16 encLen = DHKEM_X25519_ENC_LEN;
+#ifdef WOLFSSL_SMALL_STACK
+    Hpke* hpke;
+    WC_RNG* rng;
+#else
+    Hpke hpke[1];
+    WC_RNG rng[1];
+#endif
+
+    if (ctx == NULL || publicName == NULL)
+      return BAD_FUNC_ARG;
+
+    ctx->echConfigs = (WOLFSSL_EchConfig*)XMALLOC(sizeof(WOLFSSL_EchConfig),
+        ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (ctx->echConfigs == NULL)
+        return MEMORY_E;
+
+    ForceZero(ctx->echConfigs, sizeof(WOLFSSL_EchConfig));
+
+#ifdef WOLFSSL_SMALL_STACK
+    hpke = (Hpke*)XMALLOC(sizeof(Hpke), ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (hpke == NULL) {
+        XFREE(ctx->echConfigs, ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        return MEMORY_E;
+    }
+
+    rng = (WC_RNG*)XMALLOC(sizeof(Hpke), ctx->heap, DYNAMIC_TYPE_RNG);
+
+    if (rng == NULL) {
+        XFREE(ctx->echConfigs, ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        XFREE(hpke, ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        return MEMORY_E;
+    }
+#endif
+
+    ret = rngRet = wc_InitRng(rng);
+
+    /* set random config id */
+    if (rngRet == 0)
+        ret = wc_RNG_GenerateByte(rng, &ctx->echConfigs->configId);
+
+    if (ret == 0) {
+        /* set the kem id */
+        ctx->echConfigs->kemId = DHKEM_X25519_HKDF_SHA256;
+
+        /* set the cipher suite, only 1 for now */
+        ctx->echConfigs->numCipherSuites = 1;
+        ctx->echConfigs->cipherSuites =
+            (EchCipherSuite*)XMALLOC(sizeof(EchCipherSuite), ctx->heap,
+            DYNAMIC_TYPE_TMP_BUFFER);
+
+        if (ctx->echConfigs->cipherSuites == NULL) {
+            ret = MEMORY_E;
+        }
+        else {
+            ctx->echConfigs->cipherSuites[0].kdfId = HKDF_SHA256;
+            ctx->echConfigs->cipherSuites[0].aeadId = HPKE_AES_128_GCM;
+        }
+    }
+
+    if (ret == 0)
+        ret = wc_HpkeInit(hpke, DHKEM_X25519_HKDF_SHA256, HKDF_SHA256,
```

Comment:
@jpbland1 why is this hard coded for X25519? We also support ECC, so wondering how that case gets used? Is there a way to know the desired asymmetric algo at this point?

Reply:
@jpbland1 please address this comment. Otherwise the PR looks good and ready to merge by @JacobBarthelmeh .

Reply:
I just had it setup the default config that cloudflare was using, I can change it to take the algorithms the user wants but I think my rationale was wanting to make it simple

Reply:
I should also note that you can actually make an indefinite number of configs but I still think we should keep things simple and just have 1 config and add support for multiple if someone asks for it